felix/dataplane/linux/bpf_ep_mgr.go

```diff
@@ -943,7 +1232,45 @@ func (m *bpfEndpointManager) applyProgramsToDirtyDataInterfaces() {
 				"Ignoring interface that doesn't match the host data/l3 interface regex")
 			return nil
 		}
-		if !m.ifaceIsUp(iface) {
+
+		var (
+			err                           error
+			up                            bool
+			xdpIdx, ingressIdx, egressIdx int
+		)
+
+		m.ifacesLock.Lock()
+		defer m.ifacesLock.Unlock()
+		m.withIface(iface, func(iface *bpfInterface) bool {
+			up = iface.info.ifaceIsUp()
+
+			if xdpIdx = iface.dpState.policyIdx[hook.XDP]; xdpIdx == -1 {
+				if xdpIdx, err = m.xdpPolicyMapAlloc.Get(); err != nil {
+					return false
+				}
+			}
+			if ingressIdx = iface.dpState.policyIdx[hook.Ingress]; ingressIdx == -1 {
+				if ingressIdx, err = m.policyMapAlloc.Get(); err != nil {
+					return false
```

Comment:
If this one fails, the `xdpIdx` will be leaked and so on.

Reply:
that is not a leak as it will ledge an error, so it stays allocated for the device until either it is resolved or is freed when the device is released. We got here to a situation where we exhausted the map, the new device can be instantiated until another device is freed. Once that happens, we carry on. Or we need to resize the map and that is handled differently.

Reply:
True, if the map is full we're likely in trouble anyway(!) but I think the code leaks here because you don't store `iface.dpState.policyIdx[hook.XDP] = xdpIdx` until a few lines below; so the index is only in a local variable when you return from this func and it gets lost.